File: tiktok_downloader/urls.py
"""TikTok URL parsing / normalization.

Ports the URL handling from MainForm.cs:
- video/photo id extraction via regex
- short-link (vm./vt.tiktok.com) resolution by following the redirect
"""

from __future__ import annotations

import re
from dataclasses import dataclass

import requests

# Same patterns the C# app uses (MainForm.cs).
_VIDEO_ID_RE = re.compile(r"/video/(\d+)")
_PHOTO_ID_RE = re.compile(r"/photo/(\d+)")
_USERNAME_RE = re.compile(r"/@([\w.]+)")

SHORT_LINK_HOSTS = ("vm.tiktok.com", "vt.tiktok.com")

# A desktop UA is enough to follow the short-link redirect.
_REDIRECT_HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36"
    )
}


@dataclass
class TikTokLink:
    """A resolved TikTok post link."""

    url: str  # the full (redirect-resolved) URL
    media_id: str  # the numeric aweme id
    kind: str  # "video" or "photo"
    username: str | None = None  # @handle without the leading @

    @property
    def is_photo(self) -> bool:
        return self.kind == "photo"


def is_short_link(url: str) -> bool:
    return any(host in url for host in SHORT_LINK_HOSTS)


def resolve_short_link(url: str, timeout: int = 30) -> str:
    """Follow vm./vt.tiktok.com redirects and return the final URL.

    Mirrors GetRedirectUrl() in the C# app: a GET that lets requests follow
    redirects, then reads the final resolved URL.
    """
    resp = requests.get(
        url, headers=_REDIRECT_HEADERS, allow_redirects=True, timeout=timeout
    )
    return resp.url


def extract_username(url: str) -> str | None:
    match = _USERNAME_RE.search(url)
    return match.group(1) if match else None


def parse(url: str, resolve: bool = True, timeout: int = 30) -> TikTokLink:
    """Parse a TikTok post URL into a TikTokLink.

    If `resolve` is True, short links are expanded first. Raises ValueError when
    no video/photo id can be found.
    """
    url = url.strip()
    if resolve and is_short_link(url):
        url = resolve_short_link(url, timeout=timeout)

    video_match = _VIDEO_ID_RE.search(url)
    if video_match:
        return TikTokLink(
            url=url,
            media_id=video_match.group(1),
            kind="video",
            username=extract_username(url),
        )

    photo_match = _PHOTO_ID_RE.search(url)
    if photo_match:
        return TikTokLink(
            url=url,
            media_id=photo_match.group(1),
            kind="photo",
            username=extract_username(url),
        )

    raise ValueError(f"Could not find a video/photo id in URL: {url!r}")

File: backend/urls.go
// Package backend ports the TikTok download logic from the Python CLI
// (tiktok-downloader-py) to Go, for use behind the Wails desktop UI.
package backend

import (
	"net/http"
	"regexp"
	"strings"
	"time"
)

var (
	videoIDRe = regexp.MustCompile(`/video/(\d+)`)
	photoIDRe = regexp.MustCompile(`/photo/(\d+)`)
	usernRe   = regexp.MustCompile(`/@([\w.]+)`)
)

var shortLinkHosts = []string{"vm.tiktok.com", "vt.tiktok.com"}

const desktopUA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " +
	"(KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36"

// TikTokLink is a resolved TikTok post link.
type TikTokLink struct {
	URL      string `json:"url"`
	MediaID  string `json:"media_id"`
	Kind     string `json:"kind"` // "video" | "photo"
	Username string `json:"username"`
}

// IsPhoto reports whether the link points to a photo (slideshow) post.
func (l TikTokLink) IsPhoto() bool { return l.Kind == "photo" }

func isShortLink(url string) bool {
	for _, h := range shortLinkHosts {
		if strings.Contains(url, h) {
			return true
		}
	}
	return false
}

// resolveShortLink follows vm./vt.tiktok.com redirects and returns the final URL.
func resolveShortLink(url string, timeout time.Duration) (string, error) {
	client := &http.Client{Timeout: timeout}
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("User-Agent", desktopUA)
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	return resp.Request.URL.String(), nil
}

func extractUsername(url string) string {
	if m := usernRe.FindStringSubmatch(url); m != nil {
		return m[1]
	}
	return ""
}

// ParseURL parses a TikTok post URL into a TikTokLink, resolving short links.
func ParseURL(url string) (TikTokLink, error) {
	url = strings.TrimSpace(url)
	if isShortLink(url) {
		if resolved, err := resolveShortLink(url, 30*time.Second); err == nil {
			url = resolved
		}
	}

	if m := videoIDRe.FindStringSubmatch(url); m != nil {
		return TikTokLink{URL: url, MediaID: m[1], Kind: "video", Username: extractUsername(url)}, nil
	}
	if m := photoIDRe.FindStringSubmatch(url); m != nil {
		return TikTokLink{URL: url, MediaID: m[1], Kind: "photo", Username: extractUsername(url)}, nil
	}
	return TikTokLink{}, &ParseError{URL: url}
}

// ParseError signals that no video/photo id could be found in a URL.
type ParseError struct{ URL string }

func (e *ParseError) Error() string {
	return "could not find a video/photo id in URL: " + e.URL
}
